### ISSUE TITLE
Auth header included in HTTP requests

### ISSUE THREAD

**Issue by mattheworiordan** [ably/ably-java#166]
I am seeing a lot of this in the logs from the Java client:

```
16-09-27T11:53:14.235Z lvl=info msg="POST /channels/public:foo/messages" reqid=4e5cee24-def5-4383-bc97-aa2dd7f0f4ef host=52.31.254.190 proto=HTTP/1.1 referer= ua=Java/1.8.0_11 backend=rest-https secs=0.002 status=401 size=166
t=2016-09-27T11:53:14.247Z lvl=info msg="POST /channels/public:foo/messages" reqid=81688bd7-12da-4753-b572-d3bb47b9d17f host=52.31.254.190 proto=HTTP/1.1 referer= ua=Java/1.8.0_11 backend=rest-https app=27d-sg channel=foo:public:foo secs=0.007 status=204 size=0
```

It looks to me like it's plausible that the Java client does not send the Basic authentication headers in the request but instead waits for a challenge each time and resubmitting the request.  Given it is never valid to send a request without either a Token or Basic Auth header, we should look into this.

**Comment by mattheworiordan** [comment]
@trenouf any idea how the underlying HTTP lib works?

**Comment by trenouf** [comment]
No, sorry, never looked at it.

> @trenouf <https://github.com/trenouf> any idea how the underlying HTTP lib works?

**Comment by mattheworiordan** [comment]
Ok, might be one for @paddybyers to comment on later when he's back then.  Seems odd that every Java request is in fact two HTTP requests

**Comment by paddybyers** [comment]
> Seems odd that every Java request is in fact two HTTP requests

Well it's supposed to be doing non-preemptive authentication, which is generally preferred practice, and sending credentials in response to the www-authenticate challenge. This should only happen once; the credentials should be cached by the HTTP client and always re-sent for subsequent requests by the same client to the same origin. If that's not happening then it's a bug.

**Comment by mattheworiordan** [comment]
> Well it's supposed to be doing non-preemptive authentication, which is generally preferred practice, and sending credentials in response to the www-authenticate challenge

Why is that preferred practice when we know 100% of the time we'll need to authenticate?

> his should only happen once; the credentials should be cached by the HTTP client and always re-sent for subsequent requests by the same client to the same origin

I don't think we should assume that REST libraries are stateful.

**Comment by mattheworiordan** [comment]
@paddybyers any thoughts, not sure the advantage of challenge response ever when we know we need to authenticate?

**Comment by paddybyers** [comment]
I think we can easily change it to preemptively authenticate. The fact that the REST API could be statelessly used does change things. The original implementation based on Apache HttpClient didn't support preemptive auth, but now we are using HttpURLConnection it should be simple.

**Comment by mattheworiordan** [comment]
@paddybyers @psolstice whilst we're fixing other stuff, how easy would it be to make this change?  It seems incredibly wasteful to effectively make each HTTP request require two request/response operations at present.

**Comment by psolstice** [comment]
Took a look at the code, only the first request sent to the server is sent without auth info. Auth header is cached for later use. I see no problem changing it so even the first request contains auth header.

**Comment by paddybyers** [comment]
> I see no problem changing it so even the first request contains auth header.

... provided it is available.

There are calls that cannot work if the auth header is required, such as `/time` and `/requestToken`.

**Comment by psolstice** [comment]
I have already started testing it, yes, it leads to some change of code related to token request. Are there any other requests that I should be worried about (except `/time` and `/requestToken`)?

**Comment by mattheworiordan** [comment]
> I see no problem changing it so even the first request contains auth header.

I think we should change that then, I don't see any reason why we'd try and anonymous request and wait for an auth challenge when we know we need to authenticate always.

As mentioned by @paddybyers though, this is not the highest priority

**Comment by paddybyers** [comment]
This is released in 1.0.1
